### PR TITLE
MotoROS v1.8.0

### PR DESCRIPTION
**Pull request by ted-miller**
- Add support for YRC1000micro robot controller.
- Fix unit scaling for velocity feedback on DX100 and FS100.
- Improve detection of HC-10 robot to stop initialization of MotoROS.
- Rename executable (.out) file to indicate version number.

Co-Authored-By: @Tom-Pfahler

### CODE REVIEW

**Comment by ted-miller** [comment]
@Tom-Pfahler please reference this PR for your project.

**Comment by gavanderhoorn** [comment]
Overall comment (for now): adding version info to the binary filename would probably not be necessary if #193 got implemented (at least the version bit). Adding the version nrs would essentially revert what we did in #195.

**Comment by ted-miller** [comment]
I think there is definite benefit to having the version number. I agree with #196.

It will be a long while before I get time to implement #193 and I'm fine with maintaining the binary links (#195) for the time being.

File: motoman_driver/MotoPlus/Controller.h [hunk 1]
```diff
@@ -58,6 +58,12 @@
 #define MAX_MOTION_CONNECTIONS	1
 #define MAX_STATE_CONNECTIONS	4
 
+#if (DX100)
+#define MAX_CONTROLLABLE_GROUPS	2
```
Comment (andreaskoepf) [inline motoman_driver/MotoPlus/Controller.h]: Limiting the group-count to 2 for DX100 is not good as I explained in [using-the-motoman-sda20d-dual-arms-dx100-ros](https://discourse.ros.org/t/using-the-motoman-sda20d-dual-arms-dx100-ros-interface-indigo/4667/19).

Comment (ted-miller) [reply motoman_driver/MotoPlus/Controller.h]: Interesting... I need to investigate.

**Review by gavanderhoorn — DISMISSED** [review]
Let's keep this on `changes-requested` until @ted-miller figures out the situation with the DX100.

**Comment by gavanderhoorn** [comment]
If you'd like to get the other changes in @ted-miller, you could revert 39c6eb9 and we could merge.

**Comment by andreaskoepf** [comment]
I am not sure how important the group-limit for the DX100 really is in practice, e.g. it is probably not possible to actuate more than two groups on a DX100 with the original driver. Maybe the "DX100 >2 controllable groups" feature should be implemented later.
The only benefit I see from not limiting the group count is that the [state server](https://github.com/ros-industrial/motoman/blob/a87015adee079370c6ea206ad4e538120e6e1a83/motoman_driver/MotoPlus/StateServer.c#L116) also relies on `controller->numGroup` and it would still be possible to read out the position of the external axis:  As far as I remember we got our robot preconfigured by Yaskawa for ROS usage with a shared base axis (e.g. B1 & B2, with in total four groups) and while we could not move the base axis with ROS we at least got joint-feedback from all axis.

@ted-miller maybe this helps, I have stored the startup output of the Motoplus driver on our DX100 system (after removing the B2 axis): 

```
DX100 Speed-Cap is set to 1.200000%
axisType[0]: Rot	Rot	Rot	Rot	Rot	Rot	Rot	---	;
pulse->unit[0]: 58670.8789	58670.8789	65841.7656	65841.7656	65841.7656	33246.8320	65841.7656	--	;
maxInc[0] (in motoman joint order): 1670, 1670, 1874, 2205, 2205, 2227, 1874
maxSpeed[0] (in ros joint order): 3.557983, 3.557983, 3.557772, 3.557772, 4.186172, 4.186172, 8.372978
DX100 Speed-Cap is set to 1.200000%
axisType[1]: Rot	Rot	Rot	Rot	Rot	Rot	Rot	---	;
pulse->unit[1]: 58670.8789	58670.8789	65841.7656	65841.7656	65841.7656	33246.8320	65841.7656	--	;
maxInc[1] (in motoman joint order): 1670, 1670, 1874, 2205, 2205, 2227, 1874
maxSpeed[1] (in ros joint order): 3.557983, 3.557983, 3.557772, 3.557772, 4.186172, 4.186172, 8.372978
DX100 Speed-Cap is set to 0.400000%
axisType[2]: Rot	---	---	---	---	---	---	---	;
pulse->unit[2]: 58670.8789	--	--	--	--	--	--	--	;
maxInc[2] (in motoman joint order): 425, 0, 0, 0, 0, 0, 0
maxSpeed[2] (in ros joint order): 0.905475, 0.000000, 0.000000, 0.000000, 0.000000, 0.000000, 0.000000
Controller connection server running
```

(Sorry for the wrong Speed-Cap output .. obviously they are not percent but a factor.)

**Comment by ted-miller** [comment]
I just posted an update that I think will work for the DX100 SDA. However, I don't have one to test with... so this code is completely untested.

I'll try to test later this week on a simulator. But, if @andreaskoepf or @Sinai_Aranda can test on a live robot, I would certainly appreciate it.

To elaborate on "Add detection of DX100 SDA waist axis configuration."
I am detecting if groups 3 and 4 both have a single axis that is configured as a "shared axis". In this case, I raise an alarm that indicates it needs to be reconfigured.  I plan to add this alarm to the wiki and tell them to contact Yaskawa Tech Support.

**Comment by gavanderhoorn** [comment]
Do I understand things correctly here when I believe that you've essentially implemented synchronous motion across all groups for the DX100 with these changes?

**Comment by ted-miller** [comment]
I have tested the latest code on a controller in simulation-mode (DX100 SDA10). It appears to work.  The controller must be configured as two arms with a base axis on the first arm.  Using the latest changes, I can control all 15 axes.
Thanks to @andreaskoepf for the tip.

**Comment by gavanderhoorn** [comment]
Would be great if @andreaskoepf could test this as well.

**Comment by andreaskoepf** [comment]
I will ask @alemme to jump in for me (I am currently out of office).

**Comment by SinaiAranda** [comment]
@ted-miller I load MotoRosDX1_180.out for sda20d robot, I have this problem ([same Using the Motoman SDA20D Dual-arms DX100 ROS Interface](https://discourse.ros.org/t/using-the-motoman-sda20d-dual-arms-dx100-ros-interface-indigo/4667)). Could be a problem of motoman_driver, already that the joints of group 0 do not match in /joints_states, but yes they are published in /sda20d/sda20d_r1_controller/joint_states.

```
sda20@sda20-Aspire-S3:~/cidesi_robots_ws$ roslaunch motoman_sda20d_moveit_config robot_interface_streaming_sda20d.launch --screen
... logging to /home/sda20/.ros/log/ebd04320-6365-11e8-a180-74de2b86ebf7/roslaunch-sda20-Aspire-S3-6038.log
Checking log directory for disk usage. This may take awhile.
Press Ctrl-C to interrupt
Done checking log file disk usage. Usage is <1GB.

started roslaunch server http://sda20-Aspire-S3:45863/

SUMMARY
========

PARAMETERS
 * /controller_joint_names: ['arm_left_joint_...
 * /robot_ip_address: 192.168.255.10
 * /rosdistro: indigo
 * /rosversion: 1.11.21
 * /topic_list: [{'joints': ['arm...

NODES
  /
    joint_state (motoman_driver/robot_state)
    joint_trajectory_action (motoman_driver/motoman_driver_joint_trajectory_action)
    motion_streaming_interface (motoman_driver/motion_streaming_interface)

ROS_MASTER_URI=http://localhost:11311

core service [/rosout] found
process[joint_state-1]: started with pid [6056]
process[motion_streaming_interface-2]: started with pid [6057]
[ INFO] [1527615177.107739996]: Added message handler for message type: 0
process[joint_trajectory_action-3]: started with pid [6065]
[ INFO] [1527615177.115218686]: Robot state connecting to IP address: '192.168.255.10:50241'
[ INFO] [1527615177.118390135]: Loading topic list
[ INFO] [1527615177.118585897]: Found 3 topics
[ INFO] [1527615177.118723675]: Topic(state_value): [group:0,joints:{arm_left_joint_1_s,arm_left_joint_2_l,arm_left_joint_3_e,arm_left_joint_4_u,arm_left_joint_5_r,arm_left_joint_6_b,arm_left_joint_7_t},name:sda20d_r1_controller,ns:sda20d]
[ INFO] [1527615177.118851623]: Topic(state_value): [group:1,joints:{arm_right_joint_1_s,arm_right_joint_2_l,arm_right_joint_3_e,arm_right_joint_4_u,arm_right_joint_5_r,arm_right_joint_6_b,arm_right_joint_7_t},name:sda20d_r2_controller,ns:sda20d]
[ INFO] [1527615177.118979094]: Topic(state_value): [group:2,joints:{torso_joint_b1},name:sda20d_b1_controller,ns:sda20d]
[ INFO] [1527615177.119103594]: Loading group: [group:0,joints:{arm_left_joint_1_s,arm_left_joint_2_l,arm_left_joint_3_e,arm_left_joint_4_u,arm_left_joint_5_r,arm_left_joint_6_b,arm_left_joint_7_t},name:sda20d_r1_controller,ns:sda20d]
[ INFO] [1527615177.119357069]: Loading group: [group:1,joints:{arm_right_joint_1_s,arm_right_joint_2_l,arm_right_joint_3_e,arm_right_joint_4_u,arm_right_joint_5_r,arm_right_joint_6_b,arm_right_joint_7_t},name:sda20d_r2_controller,ns:sda20d]
[ INFO] [1527615177.119527826]: Loading group: [group:2,joints:{torso_joint_b1},name:sda20d_b1_controller,ns:sda20d]
[ INFO] [1527615177.119666147]: Loaded 3 groups
[ INFO] [1527615177.119817342]:  Initializing robot state 3 groups
[ INFO] [1527615177.119941453]: Initializing message manager with default comms fault handler
[ INFO] [1527615177.120049164]: Default communications fault handler successfully initialized
[ INFO] [1527615177.120152560]: Initializing message manager
[ INFO] [1527615177.120259623]: Added message handler for message type: 1
[ INFO] [1527615177.130645366]: Joint Trajectory Interface connecting to IP address: '192.168.255.10:50240'
[ INFO] [1527615177.133730213]: Loading topic list
[ INFO] [1527615177.133839365]: Found 3 topics
[ INFO] [1527615177.133926468]: Topic(state_value): [group:0,joints:{arm_left_joint_1_s,arm_left_joint_2_l,arm_left_joint_3_e,arm_left_joint_4_u,arm_left_joint_5_r,arm_left_joint_6_b,arm_left_joint_7_t},name:sda20d_r1_controller,ns:sda20d]
[ INFO] [1527615177.134000941]: Topic(state_value): [group:1,joints:{arm_right_joint_1_s,arm_right_joint_2_l,arm_right_joint_3_e,arm_right_joint_4_u,arm_right_joint_5_r,arm_right_joint_6_b,arm_right_joint_7_t},name:sda20d_r2_controller,ns:sda20d]
[ INFO] [1527615177.134069917]: Topic(state_value): [group:2,joints:{torso_joint_b1},name:sda20d_b1_controller,ns:sda20d]
[ INFO] [1527615177.134151560]: Loading group: [group:0,joints:{arm_left_joint_1_s,arm_left_joint_2_l,arm_left_joint_3_e,arm_left_joint_4_u,arm_left_joint_5_r,arm_left_joint_6_b,arm_left_joint_7_t},name:sda20d_r1_controller,ns:sda20d]
[ INFO] [1527615177.134327705]: Loading group: [group:1,joints:{arm_right_joint_1_s,arm_right_joint_2_l,arm_right_joint_3_e,arm_right_joint_4_u,arm_right_joint_5_r,arm_right_joint_6_b,arm_right_joint_7_t},name:sda20d_r2_controller,ns:sda20d]
[ INFO] [1527615177.134414091]: Loading group: [group:2,joints:{torso_joint_b1},name:sda20d_b1_controller,ns:sda20d]
[ INFO] [1527615177.134498863]: Loaded 3 groups
[ INFO] [1527615177.134579217]: MotomanJointTrajectoryStreamer: init
[ INFO] [1527615177.134640386]: JointTrajectoryStreamer: init
[ INFO] [1527615177.134999739]: Added message handler for message type: 10
[ INFO] [1527615177.135413286]: Connected to server
[ INFO] [1527615177.140887078]: Added message handler for message type: 15
[ INFO] [1527615177.141061787]: Creating joint_feedback_ex_relay_handler with 3 groups
[ WARN] [1527615177.142478719]: Unable to read velocity limits from 'robot_description' param.  Velocity validation disabled.
[ INFO] [1527615177.155725197]: Added message handler for message type: 2017
[ INFO] [1527615177.159234165]: Connected to server
[ INFO] [1527615177.159301737]: Successfully initialized robot state interface
[ INFO] [1527615177.159363876]: Entering message manager spin loop
[ INFO] [1527615177.161383550]: Loading topic list
[ INFO] [1527615177.161497380]: Found 3 topics
[ INFO] [1527615177.161564921]: Topic(state_value): [group:0,joints:{arm_left_joint_1_s,arm_left_joint_2_l,arm_left_joint_3_e,arm_left_joint_4_u,arm_left_joint_5_r,arm_left_joint_6_b,arm_left_joint_7_t},name:sda20d_r1_controller,ns:sda20d]
[ INFO] [1527615177.161642717]: Topic(state_value): [group:1,joints:{arm_right_joint_1_s,arm_right_joint_2_l,arm_right_joint_3_e,arm_right_joint_4_u,arm_right_joint_5_r,arm_right_joint_6_b,arm_right_joint_7_t},name:sda20d_r2_controller,ns:sda20d]
[ INFO] [1527615177.161700637]: Topic(state_value): [group:2,joints:{torso_joint_b1},name:sda20d_b1_controller,ns:sda20d]
[ INFO] [1527615177.161772453]: Loading group: [group:0,joints:{arm_left_joint_1_s,arm_left_joint_2_l,arm_left_joint_3_e,arm_left_joint_4_u,arm_left_joint_5_r,arm_left_joint_6_b,arm_left_joint_7_t},name:sda20d_r1_controller,ns:sda20d]
[ INFO] [1527615177.161916344]: Loading group: [group:1,joints:{arm_right_joint_1_s,arm_right_joint_2_l,arm_right_joint_3_e,arm_right_joint_4_u,arm_right_joint_5_r,arm_right_joint_6_b,arm_right_joint_7_t},name:sda20d_r2_controller,ns:sda20d]
[ INFO] [1527615177.161987739]: Loading group: [group:2,joints:{torso_joint_b1},name:sda20d_b1_controller,ns:sda20d]
[ INFO] [1527615177.162048215]: Loaded 3 groups
[ INFO] [1527615177.234910834]: Unlocking mutex
[ INFO] [1527615177.234995095]: Starting Motoman joint trajectory streamer thread
[ INFO] [1527615177.240152678]: Connecting to robot motion server
[ WARN] [1527615177.240220909]: Tried to connect when socket already in connected state
[FATAL] [1527615177.333382993]: ASSERTION FAILED
	file = /home/sda20/cidesi_robots_ws/src/motoman-indigo-devel/motoman_driver/src/industrial_robot_client/joint_relay_handler.cpp
	line = 325
	cond = all_joint_state.positions.size() == all_joint_names.size()

[joint_state-1] process has died [pid 6056, exit code -5, cmd /home/sda20/cidesi_robots_ws/devel/lib/motoman_driver/robot_state __name:=joint_state __log:=/home/sda20/.ros/log/ebd04320-6365-11e8-a180-74de2b86ebf7/joint_state-1.log].
log file: /home/sda20/.ros/log/ebd04320-6365-11e8-a180-74de2b86ebf7/joint_state-1*.log
```

**Comment by gavanderhoorn** [comment]
@alemme: do you see any chance to test this?

**Comment by gavanderhoorn** [comment]
Just tested this on our SIA20F + rail and there were no regressions.

Don't have an SDA, so I cannot verify those changes.

I'll wait one week for @alemme to respond, after which we'll merge, seeing as @ted-miller has already tested this.

**Comment by andreaskoepf** [comment]
@gavanderhoorn Unfortunately neither Andre nor me have found time for testing this PR.
It is not super-easy for us since our whole setup is driven by our fork of the driver (both motoplus und ros side with significant changes) which is always started from within Rosvita with GUI configuration etc. 

I estimate that I would need at least free 1-2h to test this PR (which includes motoplus updates etc). I would love to do this but currenty I just do not find the free time neccessary to do the tests (working on a holiday like today and in general in 60h+ mode).

**Comment by gavanderhoorn** [comment]
Ok, fair enough.

It was just that you offered testing changes to the driver on ROS Discourse, and you are one of the few people here with an SDA+DX100, so that would have been convenient.

We'll take you off the list then.

**Comment by gavanderhoorn** [comment]
I'm going to merge this as I haven't detected any regressions and we have at least one test by @ted-miller that shows `v1.8.0` to fix the issue with DX100+SDA and synchronous motion across all groups.

@sinaiaranda-CIDESI: I believe the issue you are running into is not related to this update to MotoROS, but is on the ROS side.

**Review by gavanderhoorn — APPROVED** [review]

**Comment by gavanderhoorn** [comment]
Thanks @ted-miller, @Tom-Pfahler for the development and @andreaskoepf for your input on DX100+SDA configs.

**Comment by SinaiAranda** [comment]
I changed to Ubuntu 16.04 with ROS kinetic. I have the same problem...

```
[FATAL] [1529595014.239579918]: ASSERTION FAILED
	file = /home/cidesi-nl/cidesi_ws/src/motoman-kinetic-devel/motoman_driver/src/industrial_robot_client/joint_relay_handler.cpp
	line = 320
	cond = all_joint_state.positions.size() == all_joint_names.size()

[joint_state-1] process has died [pid 8214, exit code -5, cmd /home/cidesi-nl/cidesi_ws/devel/lib/motoman_driver/robot_state __name:=joint_state __log:=/home/cidesi-nl/.ros/log/2af58e80-7567-11e8-99ff-a0a8cd822c0b/joint_state-1.log].
log file: /home/cidesi-nl/.ros/log/2af58e80-7567-11e8-99ff-a0a8cd822c0b/joint_state-1*.log

```
```
header: 
  seq: 1890
  stamp: 
    secs: 1529594994
    nsecs:  32893327
  frame_id: ''
name: [arm_left_joint_1_s, arm_left_joint_2_l, arm_left_joint_3_e, arm_left_joint_4_u, arm_left_joint_5_r,
  arm_left_joint_6_b, arm_left_joint_7_t]
position: []
velocity: []
effort: []

header: 
  seq: 1891
  stamp: 
    secs: 1529594994
    nsecs:  32942940
  frame_id: ''
name: [torso_joint_b1]
position: [-9.242676242138259e-06]
velocity: [0.0]
effort: []

header: 
  seq: 1892
  stamp: 
    secs: 1529594994
    nsecs:  33003155
  frame_id: ''
name: [arm_right_joint_1_s, arm_right_joint_2_l, arm_right_joint_3_e, arm_right_joint_4_u,
  arm_right_joint_5_r, arm_right_joint_6_b, arm_right_joint_7_t]
position: [-1.704423084447626e-05, 0.0, 0.0, 0.0, 0.0, 1.5187928511295468e-05, 0.0]
velocity: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
effort: []
```
It seems he does not read the positions of robot group 0 when YAML files are load.

**Comment by gavanderhoorn** [comment]
@sinaiaranda-CIDESI: can you please open a new issue for this? A PR is not the right place to investigate.

I'll see if I can find some time next week to take a look at it.

**Pull request merged**